src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs

```diff
@@ -0,0 +1,185 @@
+                PrimitiveType.SByte => "System.SByte",
+                PrimitiveType.Char => "System.Char",
+                PrimitiveType.Int16 => "System.Int16",
+                PrimitiveType.UInt16 => "System.UInt16",
+                PrimitiveType.Int32 => "System.Int32",
+                PrimitiveType.UInt32 => "System.UInt32",
+                PrimitiveType.Int64 => "System.Int64",
+                PrimitiveType.Single => "System.Single",
+                PrimitiveType.Double => "System.Double",
+                PrimitiveType.Decimal => "System.Decimal",
+                PrimitiveType.TimeSpan => "System.TimeSpan",
+                PrimitiveType.DateTime => "System.DateTime",
+                PrimitiveType.String => "System.String",
+                ObjectPrimitiveType => "System.Object",
+                IntPtrPrimitiveType => "System.IntPtr",
+                UIntPtrPrimitiveType => "System.UIntPtr",
+                _ => "System.UInt64",
+            };
+
+            s_PrimitiveTypeNames[(int)primitiveType] = typeName = TypeName.Parse(fullName.AsSpan()).WithCoreLibAssemblyName();
+        }
+        return typeName;
+    }
+
+    internal static TypeName GetPrimitiveSZArrayTypeName(PrimitiveType primitiveType)
+    {
+        TypeName? typeName = s_PrimitiveSZArrayTypeNames[(int)primitiveType];
+        if (typeName is null)
+        {
+            s_PrimitiveSZArrayTypeNames[(int)primitiveType] = typeName = GetPrimitiveTypeName(primitiveType).MakeSZArrayTypeName();
+        }
+        return typeName;
+    }
+
+    internal static PrimitiveType GetPrimitiveType<T>()
+    {
+        if (typeof(T) == typeof(bool)) return PrimitiveType.Boolean;
+        else if (typeof(T) == typeof(byte)) return PrimitiveType.Byte;
+        else if (typeof(T) == typeof(sbyte)) return PrimitiveType.SByte;
+        else if (typeof(T) == typeof(char)) return PrimitiveType.Char;
+        else if (typeof(T) == typeof(short)) return PrimitiveType.Int16;
+        else if (typeof(T) == typeof(ushort)) return PrimitiveType.UInt16;
+        else if (typeof(T) == typeof(int)) return PrimitiveType.Int32;
+        else if (typeof(T) == typeof(uint)) return PrimitiveType.UInt32;
+        else if (typeof(T) == typeof(long)) return PrimitiveType.Int64;
+        else if (typeof(T) == typeof(ulong)) return PrimitiveType.UInt64;
+        else if (typeof(T) == typeof(float)) return PrimitiveType.Single;
+        else if (typeof(T) == typeof(double)) return PrimitiveType.Double;
+        else if (typeof(T) == typeof(decimal)) return PrimitiveType.Decimal;
+        else if (typeof(T) == typeof(DateTime)) return PrimitiveType.DateTime;
+        else if (typeof(T) == typeof(TimeSpan)) return PrimitiveType.TimeSpan;
+        else if (typeof(T) == typeof(IntPtr)) return IntPtrPrimitiveType;
+        else if (typeof(T) == typeof(UIntPtr)) return UIntPtrPrimitiveType;
+        else
+        {
+            Debug.Assert(typeof(T) == typeof(string));
```

Comment:
Is it ever valid to return string for other T's? Can we be explicit for the string case and throw instead of asserting?